File: src/routes/organizations.ts
import { Hono } from 'hono';
import { drizzle } from 'drizzle-orm/d1';
import { eq, and, sql } from 'drizzle-orm';
import { organizations, users, checklistEntries } from '../db/schema';
import { hashPassword } from '../utils/auth';
import { authMiddleware, superAdminMiddleware } from '../middleware/auth';
import type { Env } from '../index';

const router = new Hono<{ Bindings: Env }>();

// Register new organization (public endpoint)
router.post('/register', async (c) => {
  try {
    const body = await c.req.json();

    const {
      organization_name,
      organization_slug,
      contact_email,
      contact_phone,
      institution_type,
      country,
      city,
      address,
      admin_username,
      admin_password,
      admin_full_name,
    } = body;

    // Validaciones
    if (!organization_name || !organization_slug || !contact_email || !admin_username || !admin_password) {
      return c.json({ error: 'Missing required fields' }, 400);
    }

    if (admin_password.length < 8) {
      return c.json({ error: 'Password must be at least 8 characters' }, 400);
    }

    const db = drizzle(c.env.DB);

    // Verificar que el slug no exista
    const existingOrg = await db.select().from(organizations)
      .where(eq(organizations.slug, organization_slug))
      .get();

    if (existingOrg) {
      return c.json({ error: 'Organization slug already exists' }, 409);
    }

    // Verificar que el username no exista
    const existingUser = await db.select().from(users)
      .where(eq(users.username, admin_username))
      .get();

    if (existingUser) {
      return c.json({ error: 'Username already exists' }, 409);
    }

    // Crear organización
    const trialEndsAt = new Date();
    trialEndsAt.setDate(trialEndsAt.getDate() + 30);

    const newOrg = await db.insert(organizations).values({
      name: organization_name,
      slug: organization_slug,
      contactEmail: contact_email,
      contactPhone: contact_phone,
      institutionType: institution_type,
      country: country,
      city: city,
      address: address,
      isActive: true,
      plan: 'free',
      maxUsers: 5,
      trialEndsAt: trialEndsAt.toISOString(),
    }).returning().get();

    // Crear usuario admin de la organización
    const hashedPwd = await hashPassword(admin_password);

    const newUser = await db.insert(users).values({
      email: contact_email,
      username: admin_username,
      hashedPassword: hashedPwd,
      fullName: admin_full_name || admin_username,
      isActive: true,
      isAdmin: true,
      isSuperAdmin: false,
      organizationId: newOrg.id,
    }).returning().get();

    return c.json({
      message: 'Organization registered successfully',
      organization: {
        id: newOrg.id,
        name: newOrg.name,
        slug: newOrg.slug,
        plan: newOrg.plan,
        trial_ends_at: newOrg.trialEndsAt,
      },
      admin_user: {
        id: newUser.id,
        username: newUser.username,
        email: newUser.email,
      },
    }, 201);

  } catch (error) {
    console.error('Organization registration error:', error);
    return c.json({ error: 'Failed to register organization' }, 500);
  }
});

// List all organizations (super admin only)
router.get('/list', authMiddleware, superAdminMiddleware, async (c) => {
  try {
    const db = drizzle(c.env.DB);

    // Get all organizations with user count
    const orgs = await db.select({
      id: organizations.id,
      name: organizations.name,
      slug: organizations.slug,
      contactEmail: organizations.contactEmail,
      contactPhone: organizations.contactPhone,
      institutionType: organizations.institutionType,
      country: organizations.country,
      city: organizations.city,
      isActive: organizations.isActive,
      plan: organizations.plan,
      maxUsers: organizations.maxUsers,
      createdAt: organizations.createdAt,
      trialEndsAt: organizations.trialEndsAt,
      userCount: sql<number>`(SELECT COUNT(*) FROM users WHERE organization_id = ${organizations.id})`,
      activeUserCount: sql<number>`(SELECT COUNT(*) FROM users WHERE organization_id = ${organizations.id} AND is_active = 1)`,
      checklistCount: sql<number>`(SELECT COUNT(*) FROM checklist_entries WHERE organization_id = ${organizations.id})`,
    }).from(organizations).all();

    return c.json({
      total: orgs.length,
      organizations: orgs,
    });

  } catch (error) {
    console.error('List organizations error:', error);
    return c.json({ error: 'Failed to list organizations' }, 500);
  }
});

// Get organization details
router.get('/:id', authMiddleware, async (c) => {
  try {
    const orgId = parseInt(c.req.param('id'));
    const currentUser = c.get('user');

    // Los usuarios normales solo pueden ver su propia organización
    if (!currentUser.isSuperAdmin && currentUser.organizationId !== orgId) {
      return c.json({ error: 'Forbidden' }, 403);
    }

    const db = drizzle(c.env.DB);
    const org = await db.select().from(organizations)
      .where(eq(organizations.id, orgId))
      .get();

    if (!org) {
      return c.json({ error: 'Organization not found' }, 404);
    }

    return c.json(org);

  } catch (error) {
    console.error('Get organization error:', error);
    return c.json({ error: 'Failed to get organization' }, 500);
  }
});

// Toggle organization active status (super admin only)
router.post('/:id/toggle-active', authMiddleware, superAdminMiddleware, async (c) => {
  try {
    const orgId = parseInt(c.req.param('id'));
    const db = drizzle(c.env.DB);

    const org = await db.select().from(organizations)
      .where(eq(organizations.id, orgId))
      .get();

    if (!org) {
      return c.json({ error: 'Organization not found' }, 404);
    }

    const newStatus = !org.isActive;

    await db.update(organizations)
      .set({ isActive: newStatus })
      .where(eq(organizations.id, orgId))
      .run();

    return c.json({
      message: `Organization ${newStatus ? 'activated' : 'deactivated'}`,
      is_active: newStatus,
    });

  } catch (error) {
    console.error('Toggle organization error:', error);
    return c.json({ error: 'Failed to toggle organization status' }, 500);
  }
});

export default router;

File: app/routers/organizations.py
"""
Router para gestión de organizaciones y registro multi-tenant
"""
from fastapi import APIRouter, Depends, HTTPException, Request
from fastapi.responses import HTMLResponse
from fastapi.templating import Jinja2Templates
from sqlalchemy.orm import Session
from sqlalchemy import func
from datetime import datetime, timedelta
import re

from app.db.database import get_db
from app.models.organization import Organization
from app.models.user import User
from app.models.organization_schemas import OrganizationCreate, OrganizationResponse, OrganizationStats
from app.auth.users import get_password_hash, get_current_user

router = APIRouter(prefix="/organizations", tags=["organizations"])
templates = Jinja2Templates(directory="templates")

def generate_slug(name: str) -> str:
    """Genera un slug único a partir del nombre"""
    # Convertir a minúsculas y reemplazar espacios
    slug = name.lower().strip()
    slug = re.sub(r'[^\w\s-]', '', slug)  # Remover caracteres especiales
    slug = re.sub(r'[-\s]+', '-', slug)  # Reemplazar espacios con guiones
    return slug[:100]  # Limitar longitud

@router.get("/register", response_class=HTMLResponse)
async def show_register_form(request: Request):
    """Mostrar formulario de registro de organización"""
    return templates.TemplateResponse("organization_register.html", {
        "request": request,
        "title": "Registrar Organización"
    })

@router.post("/register", response_model=OrganizationResponse)
async def register_organization(org_data: OrganizationCreate, db: Session = Depends(get_db)):
    """
    Registrar una nueva organización con su administrador
    """
    # Verificar que no exista el email
    existing_user = db.query(User).filter(User.email == org_data.admin_email).first()
    if existing_user:
        raise HTTPException(
            status_code=400,
            detail="Ya existe un usuario con ese email"
        )
    
    # Generar slug único
    base_slug = generate_slug(org_data.name)
    slug = base_slug
    counter = 1
    while db.query(Organization).filter(Organization.slug == slug).first():
        slug = f"{base_slug}-{counter}"
        counter += 1
    
    # Crear organización
    trial_end = datetime.utcnow() + timedelta(days=30)  # 30 días de trial
    new_org = Organization(
        name=org_data.name,
        slug=slug,
        contact_email=org_data.contact_email,
        contact_phone=org_data.contact_phone,
        institution_type=org_data.institution_type,
        country=org_data.country,
        city=org_data.city,
        address=org_data.address,
        is_active=True,
        plan="free",
        max_users=5,
        trial_ends_at=trial_end
    )
    
    db.add(new_org)
    db.flush()  # Para obtener el ID
    
    # Crear usuario administrador
    hashed_password = get_password_hash(org_data.admin_password)
    admin_user = User(
        email=org_data.admin_email,
        username=org_data.admin_email.split('@')[0],  # Usar parte antes del @
        hashed_password=hashed_password,
        full_name=org_data.admin_name,
        is_active=True,
        is_admin=True,
        is_super_admin=False,
        organization_id=new_org.id
    )
    
    db.add(admin_user)
    db.commit()
    db.refresh(new_org)
    
    return new_org

@router.get("/list", response_class=HTMLResponse)
async def list_organizations(
    request: Request,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Panel de administración de organizaciones (solo super admin)"""
    if not current_user.is_super_admin:
        raise HTTPException(status_code=403, detail="Acceso denegado")
    
    # Obtener todas las organizaciones con estadísticas
    orgs = db.query(Organization).order_by(Organization.created_at.desc()).all()
    
    org_stats = []
    for org in orgs:
        # Contar usuarios
        total_users = db.query(func.count(User.id)).filter(
            User.organization_id == org.id
        ).scalar() or 0
        
        active_users = db.query(func.count(User.id)).filter(
            User.organization_id == org.id,
            User.is_active == True
        ).scalar() or 0
        
        # Contar checklists (importar modelo si existe)
        try:
            from app.models.checklist_entry import ChecklistEntrySQL
            total_checklists = db.query(func.count(ChecklistEntrySQL.id)).filter(
                ChecklistEntrySQL.organization_id == org.id
            ).scalar() or 0
        except:
            total_checklists = 0
        
        org_stats.append({
            "id": org.id,
            "name": org.name,
            "slug": org.slug,
            "contact_email": org.contact_email,
            "institution_type": org.institution_type,
            "country": org.country,
            "city": org.city,
            "plan": org.plan,
            "is_active": org.is_active,
            "created_at": org.created_at,
            "trial_ends_at": org.trial_ends_at,
            "total_users": total_users,
            "active_users": active_users,
            "total_checklists": total_checklists
        })
    
    return templates.TemplateResponse("super_admin_dashboard.html", {
        "request": request,
        "title": "Panel de Super Admin",
        "organizations": org_stats,
        "current_user": current_user
    })

@router.post("/{org_id}/toggle-active")
async def toggle_organization_active(
    org_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Activar/desactivar una organización"""
    if not current_user.is_super_admin:
        raise HTTPException(status_code=403, detail="Acceso denegado")
    
    org = db.query(Organization).filter(Organization.id == org_id).first()
    if not org:
        raise HTTPException(status_code=404, detail="Organización no encontrada")
    
    org.is_active = not org.is_active
    db.commit()
    
    return {"success": True, "is_active": org.is_active}

@router.get("/api/stats")
async def get_platform_stats(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Estadísticas generales de la plataforma"""
    if not current_user.is_super_admin:
        raise HTTPException(status_code=403, detail="Acceso denegado")
    
    total_orgs = db.query(func.count(Organization.id)).scalar() or 0
    active_orgs = db.query(func.count(Organization.id)).filter(Organization.is_active == True).scalar() or 0
    total_users = db.query(func.count(User.id)).scalar() or 0
    
    return {
        "total_organizations": total_orgs,
        "active_organizations": active_orgs,
        "total_users": total_users,
        "timestamp": datetime.utcnow()
    }
